### PR TITLE
[SPARK-32044][SS][2.4] Kakfa continuous processing print mislead initial offset

### PR DESCRIPTION
**Pull request by warrenzhu25**
### What changes were proposed in this pull request?

Use `java.util.Optional.orElseGet` instead of `java.util.Optional.orElse` to fix unnecessary kafka offset fetch and misleading info log. In Java, `orElseGet` uses lazy evaluation while `orElse` always evaluate the expression. 

### Why are the changes needed?

Fix mislead initial offsets log and unnecessary kafka offset fetch

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Currently, no test for KafkaContinuousReader. Also it's hard to test log.

### CODE REVIEW

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala [hunk 1]
```diff
@@ -70,15 +71,17 @@ class KafkaContinuousReader(
 
   private var offset: Offset = _
   override def setStartOffset(start: ju.Optional[Offset]): Unit = {
-    offset = start.orElse {
-      val offsets = initialOffsets match {
-        case EarliestOffsetRangeLimit => KafkaSourceOffset(offsetReader.fetchEarliestOffsets())
-        case LatestOffsetRangeLimit => KafkaSourceOffset(offsetReader.fetchLatestOffsets(None))
-        case SpecificOffsetRangeLimit(p) => offsetReader.fetchSpecificOffsets(p, reportDataLoss)
+    offset = start.orElseGet(new Supplier[Offset] {
```
Comment (HeartSaVioR) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala:74]: Nice finding! It's quite confusing on Scala-friendly developers.

**Review by HeartSaVioR — APPROVED** [review]
LGTM

**Comment by HeartSaVioR** [comment]
Probably, it would be nice to explain the difference between Optional.orElse vs Optional.orElseGet for Java a bit in PR description. It's very confusing in Scala developers as we expect lazy evaluation natively.

And one more, please rebase the target branch to master. We normally receive patches against master branch. If the problem no longer exists in master branch, we may want to find and port back the commit which fixed the issue previously, instead of picking up newer commit.

**Comment by warrenzhu25** [comment]
> Probably, it would be nice to explain the difference between Optional.orElse vs Optional.orElseGet for Java a bit in PR description. It's very confusing in Scala developers as we expect lazy evaluation natively.
> 
> And one more, please rebase the target branch to master. We normally receive patches against master branch. If the problem no longer exists in master branch, we may want to find and port back the commit which fixed the issue previously, instead of picking up newer commit.

1. Updated description to explain more about Optional.orElse vs Optional.orElseGet
2. In master branch, KafkaContinuousReader.scala is no longer existed. This class only in branch-2.4. Any suggestions about how should I do?

**Comment by HeartSaVioR** [comment]
Oh OK. Let's hear committers' voice here. Before that, can we add `[2.4]` after `[SS]` so that the target version is clear?

cc. @zsxwing

**Comment by dongjoon-hyun** [comment]
ok to test

**Review by dongjoon-hyun — APPROVED** [review]
Thank you for your first contribution and welcome, @warrenzhu25 .
The bug seems to exist only in `branch-2.4`. So, +1, LGTM.

**Comment by dongjoon-hyun** [comment]
Merged to `branch-2.4`. SPARK-32044 is assigned to you, @warrenzhu25 .
Also, thanks, @HeartSaVioR .

**Pull request closed without merging**